src/shims/native_lib/ffi.rs

```diff
@@ -0,0 +1,113 @@
+use libffi::low::CodePtr;
+use libffi::middle::{Arg as ArgPtr, Cif, Type as FfiType};
+
+/// Perform the actual FFI call.
+///
+/// SAFETY: The `FfiArg`s passed must have been correctly instantiated (i.e. their
+/// type layout must match the data they point to), and the safety invariants of
+/// the foreign function being called must be upheld (if any).
+pub unsafe fn call<'a, R: libffi::high::CType>(fun: CodePtr, args: Vec<FfiArg<'a>>) -> R {
+    let mut arg_tys = vec![];
+    let mut arg_ptrs = vec![];
+    for arg in args {
+        arg_tys.push(arg.ty);
+        arg_ptrs.push(arg.ptr)
+    }
+    let cif = Cif::new(arg_tys, R::reify().into_middle());
+    unsafe { cif.call(fun, &arg_ptrs) }
+}
+
+/// A wrapper type for `libffi::middle::Type` which also holds a pointer to the data.
+pub struct FfiArg<'a> {
+    /// The type layout information for the pointed-to data.
+    ty: FfiType,
+    /// A pointer to the data described in `ty`.
+    ptr: ArgPtr,
+    /// Lifetime of the actual pointed-to data.
+    _p: std::marker::PhantomData<&'a [u8]>,
+}
+
+impl<'a> FfiArg<'a> {
+    fn new(ty: FfiType, ptr: ArgPtr) -> Self {
+        Self { ty, ptr, _p: std::marker::PhantomData }
+    }
+}
+
+/// An owning form of `FfiArg`.
+/// We introduce this enum instead of just calling `Arg::new` and storing a list of
+/// `libffi::middle::Arg` directly, because the `libffi::middle::Arg` just wraps a reference to
+/// the value it represents and we need to store a copy of the value, and pass a reference to
+/// this copy to C instead.
+#[derive(Debug, Clone)]
+pub enum CArg {
+    /// Primitive type.
+    Primitive(CPrimitive),
+    /// Struct with its computed type layout and bytes.
+    Struct(FfiType, Box<[u8]>),
```

Comment:
This doesn't have to be a struct, right? It can be any type that `FfiType` can represent, which includes basic integers.

Reply:
I'll change the name, presumably the next PR idea which you brought up on Zulip would make this enum redundant anyway